pkg/controller/validationscache.go

```diff
@@ -64,35 +72,45 @@ func (vc *validationCache) store(obj client.Object, outcome validations.Validati
 	)
 }
 
+// drain overwrites the current instance with an empty one
+// all keys and values will be lost
 func (vc *validationCache) drain() {
 	*vc = validationCache{}
 }
 
+// remove deletes a key, and its value, from the instance
+// it uses given object to search for the validationKey
 func (vc *validationCache) remove(obj client.Object) {
 	key := newValidationKey(obj)
 	vc.removeKey(key)
 }
 
+// removeKey deletes a key, and its value, from the instance
 func (vc *validationCache) removeKey(key validationKey) {
 	delete(*vc, key)
 }
 
+// retrieve returns the value, if exists, within a key
+// it uses given object to search for the validationKey
+// it returns a second parameter to check if the key existed in the instance
 func (vc *validationCache) retrieve(obj client.Object) (validationResource, bool) {
 	key := newValidationKey(obj)
-	val, ok := (*vc)[key]
-	return val, ok
+	val, exists := (*vc)[key]
+	return val, exists
 }
 
+// objectAlreadyValidated returns if the given object has been validated
 func (vc *validationCache) objectAlreadyValidated(obj client.Object) bool {
-	validationOutcome, ok := vc.retrieve(obj)
-	storedResourceVersion := validationOutcome.version
-	if !ok {
-		return false
+	validationOutcome, exists := vc.retrieve(obj)
+	if exists {
+		storedResourceVersion := string(validationOutcome.version)
+		currentResourceVersion := obj.GetResourceVersion()
+
+		if storedResourceVersion != currentResourceVersion {
+			vc.remove(obj)
+			return false
+		}
 	}
-	currentResourceVersion := obj.GetResourceVersion()
-	if string(storedResourceVersion) != currentResourceVersion {
-		vc.remove(obj)
-		return false
-	}
-	return true
+
+	return exists
```

Comment:
This refactor increases nesting depth for the active code path reducing clarity. The 'guard' approach before this change is preferred in this project so this should be reverted.